tools/rostopic/src/rostopic/__init__.py

```diff
@@ -173,40 +176,122 @@ def print_hz(self):
+        """
+        if not self.times:
+            return
+        ret = self.get_hz()
+        if ret is None:
+            print("no new messages")
+            return
+        rate, min_delta, max_delta, std_dev, window = ret
+        print("average rate: %.3f\n\tmin: %.3fs max: %.3fs std dev: %.5fs window: %s"%(rate, min_delta, max_delta, std_dev, window))
 
 def _sleep(duration):
     rospy.rostime.wallsleep(duration)
 
-def _rostopic_hz(topic, window_size=-1, filter_expr=None):
+
+def get_parent_topics(topic, parent_topics=None, impl=False):
+    master = rosgraph.Master('/rostopic')
+    publications, subscriptions, _ = master.getSystemState()
+    if parent_topics is None:
+        parent_topics = []
+    pub_nodes = [l for t, l in publications if t == topic]
+    if not pub_nodes:
+        return parent_topics
+    for node in pub_nodes[0]:
+        subs = [t for t, l in subscriptions if node in l]
+        subs = filter(lambda x: x not in parent_topics, subs)  # get unique
+        parent_topics.extend(subs)
+        if not impl:
+            return parent_topics
+        for sub in subs:
+            parent_topics = get_parent_topics(sub, parent_topics, impl=impl)
+    return parent_topics
+
+
+def _rostopic_hz(topics, window_size=-1, filter_expr=None, search_parent=False):
     """
     Periodically print the publishing rate of a topic to console until
     shutdown
-    :param topic: topic name, ``str``
+    :param topics: topic names, ``list`` of ``str``
     :param window_size: number of messages to average over, -1 for infinite, ``int``
     :param filter_expr: Python filter expression that is called with m, the message instance
+    :param search_parent: whether search parent topics to check the hz
     """
-    msg_class, real_topic, _ = get_topic_class(topic, blocking=True) #pause hz until topic is published
+    msg_classes, real_topics = [], []
+    for topic in topics:
+        msg_class, real_topic, _ = get_topic_class(topic, blocking=True) #pause hz until topic is published
+        msg_classes.append(msg_class)
+        real_topics.append(real_topic)
     if rospy.is_shutdown():
         return
     rospy.init_node(NAME, anonymous=True)
-    rt = ROSTopicHz(window_size, filter_expr=filter_expr)
-    # we use a large buffer size as we don't know what sort of messages we're dealing with.
-    # may parameterize this in the future
-    if filter_expr is not None:
-        # have to subscribe with topic_type
-        sub = rospy.Subscriber(real_topic, msg_class, rt.callback_hz)
+    hz_checkers = []
+    for topic in topics:
+        rt = ROSTopicHz(window_size, filter_expr=filter_expr)
+        # we use a large buffer size as we don't know what sort of messages we're dealing with.
+        # may parameterize this in the future
+        if filter_expr is not None:
+            # have to subscribe with topic_type
+            sub = rospy.Subscriber(real_topic, msg_class, rt.callback_hz)
```

Comment:
`msg_class` is only set in the previous loop (line 230) which doesn't seem to be what you want?